File: cli/src/preflights/preflight-init.ts
import chalk from "chalk";
import { join } from "path";
import { isAlreadyInitialized, validateFlutterProject } from "../utils/validator.js";

export interface PreflightResult {
  success: boolean;
  error?: string;
  warning?: string;
  details?: string[];
}

/**
 * Pre-flight checks for the init command
 * @param cwd - Current working directory
 * @returns PreflightResult with validation status and messages
 */
export function preflightInit(cwd: string): PreflightResult {
  // Step 1: Validate Flutter project
  const validation = validateFlutterProject(cwd);

  if (!validation.isValid) {
    return {
      success: false,
      error: chalk.red("✗ Flutter project validation failed"),
      details: [chalk.gray(`  • ${validation.error}`)],
    };
  }

  // Step 2: Check if already initialized (warning, not error)
  if (isAlreadyInitialized(cwd)) {
    const themePath = join(cwd, "lib", "config", "theme.dart");
    return {
      success: true,
      warning: chalk.yellow("⚠ fluttercn is already initialized"),
      details: [
        chalk.gray(`  Theme file exists at: ${themePath}`),
        chalk.gray("  Skipping initialization..."),
      ],
    };
  }

  return {
    success: true,
  };
}

File: dart-cli/lib/preflights/preflight_init.dart
import 'package:fluttercn/utils/validator.dart';
import 'package:path/path.dart' as path;

/// Preflight result with success status and optional messages
class PreflightResult {
  final bool success;
  final String? error;
  final String? warning;
  final List<String>? details;

  PreflightResult({
    required this.success,
    this.error,
    this.warning,
    this.details,
  });
}

PreflightResult preflightInit(String cwd) {
  final validation = validateFlutterProject(cwd);

  if (!validation.isValid) {
    return PreflightResult(
      success: false,
      error: '✗ Flutter project validation failed',
      details: ['  • ${validation.error}'],
    );
  }

  if (isAlreadyInitialized(cwd)) {
    final themePath = path.join(cwd, 'lib', 'config', 'theme.dart');
    return PreflightResult(
      success: true,
      warning: '⚠ fluttercn is already initialized',
      details: [
        '  Theme file exists at: $themePath',
        '  Skipping initialization...',
      ],
    );
  }

  return PreflightResult(success: true);
}
